File: harness/ppaz_inventory.mjs
#!/usr/bin/env node
// Inventory of P. Paz's Strava history export (data/activities/strava_ppaz/, gitignored:
// third-party GPS shared with consent — see the article's Ethics section).
//
// Goal: find which activities are usable for the second-rider model verification —
// i.e. CYCLING rides WITH POWER (the empirical ∫P·dt benchmark needs a power meter).
// Scans every .fit / .fit.gz, reads sport + record stats, and writes
// strava_ppaz_manifest.json (gitignored via data/activities/*.json).
//
//   node ppaz_inventory.mjs
//
// parseFIT is the verbatim copy used by all harnesses (censo_compare.mjs), extended
// ONLY to also capture the sport enum (message 12 field 0 / session 18 field 5) —
// the record decoding is untouched.
import fs from 'node:fs';
import path from 'node:path';
import zlib from 'node:zlib';
import { fileURLToPath } from 'node:url';

const HERE = path.dirname(fileURLToPath(import.meta.url));
const DATA = path.join(HERE, '..', 'data', 'activities');
const DIR = path.join(DATA, 'strava_ppaz');
const SPORT = { 0: 'generic', 1: 'run', 2: 'ride', 5: 'swim', 11: 'walk', 17: 'hike' };

function parseFIT(buffer) {
  const dv = new DataView(buffer);
  if (buffer.byteLength < 14) throw new Error('FIT muito curto');
  const headerSize = dv.getUint8(0), dataSize = dv.getUint32(4, true);
  if (String.fromCharCode(dv.getUint8(8), dv.getUint8(9), dv.getUint8(10), dv.getUint8(11)) !== '.FIT') throw new Error('no .FIT');
  const end = Math.min(headerSize + dataSize, buffer.byteLength);
  let pos = headerSize; const defs = {}, records = [];
  let lastTs;   // running timestamp for compressed-timestamp headers (5-bit offset, 32 s rollover)
  let sport;    // first sport enum seen (msg 12 field 0, or session msg 18 field 5)
  const read = (p, bt, little) => { switch (bt & 0x1F) {
    case 0x01: { const v = dv.getInt8(p); return v === 0x7F ? undefined : v; }
    case 0x00: case 0x02: case 0x0A: case 0x0D: { const v = dv.getUint8(p); return v === 0xFF ? undefined : v; }
    case 0x03: { const v = dv.getInt16(p, little); return v === 0x7FFF ? undefined : v; }
    case 0x04: case 0x0B: { const v = dv.getUint16(p, little); return v === 0xFFFF ? undefined : v; }
    case 0x05: { const v = dv.getInt32(p, little); return v === 0x7FFFFFFF ? undefined : v; }
    case 0x06: case 0x0C: { const v = dv.getUint32(p, little) >>> 0; return v === 0xFFFFFFFF ? undefined : v; }
    case 0x08: return dv.getFloat32(p, little);
    case 0x09: return dv.getFloat64(p, little);
    default: return undefined; } };
  while (pos < end) {
    const rh = dv.getUint8(pos); pos++;
    let local, isDef = false, hasDev = false, tsOffset;
    if (rh & 0x80) { local = (rh >> 5) & 0x03; tsOffset = rh & 0x1F; }
    else { local = rh & 0x0F; isDef = !!(rh & 0x40); hasDev = !!(rh & 0x20); }
    if (isDef) {
      pos++; const little = dv.getUint8(pos) === 0; pos++;
      const gmn = dv.getUint16(pos, little); pos += 2;
      const nf = dv.getUint8(pos); pos++;
      const fields = [];
      for (let i = 0; i < nf; i++) { fields.push({ num: dv.getUint8(pos), size: dv.getUint8(pos + 1), bt: dv.getUint8(pos + 2) }); pos += 3; }
      let devSize = 0;
      if (hasDev) { const nd = dv.getUint8(pos); pos++; for (let i = 0; i < nd; i++) { devSize += dv.getUint8(pos + 1); pos += 3; } }
      defs[local] = { gmn, little, fields, devSize };
    } else {
      const def = defs[local];
      if (!def) throw new Error('FIT corrompido (dado sem definição)');
      let p = pos; const rec = {};
      for (const f of def.fields) {
        if (def.gmn === 20) { const v = read(p, f.bt, def.little);
          if (v !== undefined) {
            if (f.num === 0) rec.lat = v * (180 / 2147483648);
            else if (f.num === 1) rec.lon = v * (180 / 2147483648);
            else if (f.num === 2) { if (rec.alt === undefined) rec.alt = v / 5 - 500; }
            else if (f.num === 78) rec.alt = v / 5 - 500;
            else if (f.num === 5) rec.dist = v / 100;
            else if (f.num === 6) { if (rec.speed === undefined) rec.speed = v / 1000; }
            else if (f.num === 73) rec.speed = v / 1000;
            else if (f.num === 7) rec.power = v;
            else if (f.num === 4) rec.cad = v;          // cadence (rpm) — 0 ⇒ not pedalling
            else if (f.num === 253) rec.time = v;
          } }
        else if ((def.gmn === 12 && f.num === 0) || (def.gmn === 18 && f.num === 5)) {
          const v = read(p, f.bt, def.little);
          if (v !== undefined && sport === undefined) sport = v;
        }
        else if (f.num === 253) {   // any message's timestamp advances the running clock
          const v = read(p, f.bt, def.little);
          if (v !== undefined) rec.time = v;
        }
        p += f.size;
      }
      pos = p + def.devSize;
      if (tsOffset !== undefined && rec.time === undefined && lastTs !== undefined) {
        let ts = (lastTs & ~31) | tsOffset;
        if (ts < lastTs) ts += 32;
        rec.time = ts;
      }
      if (rec.time !== undefined) lastTs = rec.time;
      if (def.gmn === 20) records.push(rec);
    }
  }
  return { records, sport };
}

const slice = b => b.buffer.slice(b.byteOffset, b.byteOffset + b.byteLength);
const FIT_EPOCH = 631065600;   // 1989-12-31 UTC, seconds

const files = fs.readdirSync(DIR).filter(f => f.endsWith('.fit') || f.endsWith('.fit.gz')).sort();
const out = []; let errors = 0;
for (const f of files) {
  try {
    let buf = fs.readFileSync(path.join(DIR, f));
    if (f.endsWith('.gz')) buf = zlib.gunzipSync(buf);
    const { records: recs, sport } = parseFIT(slice(buf));
    if (recs.length < 2) throw new Error('sem registros');
    const n = recs.length;
    let nPow = 0, nPowPos = 0, nCad = 0, nAlt = 0, maxDist = 0, t0, t1;
    for (const r of recs) {
      if (r.power !== undefined) { nPow++; if (r.power > 0) nPowPos++; }
      if (r.cad !== undefined) nCad++;
      if (r.alt !== undefined) nAlt++;
      if (r.dist !== undefined && r.dist > maxDist) maxDist = r.dist;
      if (r.time !== undefined) { if (t0 === undefined) t0 = r.time; t1 = r.time; }
    }
    out.push({
      id: f.replace(/\.fit(\.gz)?$/, ''), file: path.join('strava_ppaz', f),
      sport: SPORT[sport] ?? (sport === undefined ? 'unknown' : `sport${sport}`),
      date: t0 !== undefined ? new Date((t0 + FIT_EPOCH) * 1000).toISOString().slice(0, 10) : null,
      hours: t0 !== undefined ? +((t1 - t0) / 3600).toFixed(2) : null,
      km: +(maxDist / 1000).toFixed(1), n,
      powCov: +(nPow / n).toFixed(3), powPos: +(nPowPos / n).toFixed(3),
      cadCov: +(nCad / n).toFixed(3), altCov: +(nAlt / n).toFixed(3),
    });
  } catch (e) { errors++; }
}
fs.writeFileSync(path.join(DATA, 'strava_ppaz_manifest.json'), JSON.stringify(out, null, 1));

// ---- summary ----
const bySport = {};
for (const a of out) (bySport[a.sport] ??= []).push(a);
console.log(`P. PAZ STRAVA EXPORT — ${files.length} FIT files, ${out.length} parsed, ${errors} errors\n`);
console.log('sport      n     w/ power(>50% cov)   rides>20km w/ power');
for (const [s, arr] of Object.entries(bySport).sort((a, b) => b[1].length - a[1].length)) {
  const pow = arr.filter(a => a.powCov > 0.5);
  const big = pow.filter(a => a.km > 20);
  console.log(`${s.padEnd(9)}${String(arr.length).padStart(5)}${String(pow.length).padStart(13)}${String(big.length).padStart(19)}`);
}
const rides = (bySport.ride || []).filter(a => a.powCov > 0.5);
if (rides.length) {
  const dates = rides.map(r => r.date).filter(Boolean).sort();
  const kms = rides.map(r => r.km).sort((a, b) => a - b);
  const q = p => kms[Math.floor(p * (kms.length - 1))];
  console.log(`\nRIDES WITH POWER: ${rides.length}`);
  console.log(`  dates ${dates[0]} … ${dates[dates.length - 1]}`);
  console.log(`  km: min ${q(0)}  p25 ${q(.25)}  median ${q(.5)}  p75 ${q(.75)}  max ${q(1)}`);
  console.log(`  alt coverage ≥99%: ${rides.filter(r => r.altCov >= 0.99).length}`);
  console.log(`  cadence coverage ≥50%: ${rides.filter(r => r.cadCov >= 0.5).length}`);
}
console.log('\nwrote strava_ppaz_manifest.json');

File: harness/ppaz_inventory.py
#!/usr/bin/env python3
"""Inventory of P. Paz's Strava history export — Python port of
harness/ppaz_inventory.mjs (same console report, byte-identical manifest).
(data/activities/strava_ppaz/, gitignored: third-party GPS shared with
consent — see the article's Ethics section.)

Goal: find which activities are usable for the second-rider model verification —
i.e. CYCLING rides WITH POWER (the empirical ∫P·dt benchmark needs a power meter).
Scans every .fit / .fit.gz, reads sport + record stats, and writes
strava_ppaz_manifest.json (gitignored via data/activities/*.json).

  python3 harness/ppaz_inventory.py

parseFIT is the verbatim copy used by all harnesses (censo_compare.mjs), extended
ONLY to also capture the sport enum (message 12 field 0 / session 18 field 5) —
the record decoding is untouched. It is therefore ported locally (the bem copy
has neither the cadence field nor the sport capture).
"""

import gzip
import math
import os
import re
import struct
import sys
from datetime import datetime, timezone

HERE = os.path.dirname(os.path.abspath(__file__))
REPO = os.path.dirname(HERE)
sys.path.insert(0, os.path.join(REPO, "analysis"))

from bem.jsfmt import js_str, to_fixed

DATA = os.path.join(REPO, "data", "activities")
DIR = os.path.join(DATA, "strava_ppaz")
SPORT = {0: "generic", 1: "run", 2: "ride", 5: "swim", 11: "walk", 17: "hike"}


# ---- FIT parsing — the .mjs's EXTENDED copy (cadence field + sport enum) ----

def _reader(buf, little):
    """read(p, bt) closure — FIT base-type reads honouring the definition's
    endianness and the FIT invalid-value markers. Out-of-range reads raise
    (struct.error / IndexError), as DataView throws RangeError in JS."""
    e = "<" if little else ">"

    def read(p, bt):
        b = bt & 0x1F
        if b == 0x01:
            v = struct.unpack_from(e + "b", buf, p)[0]
            return None if v == 0x7F else v
        if b in (0x00, 0x02, 0x0A, 0x0D):
            v = buf[p]
            return None if v == 0xFF else v
        if b == 0x03:
            v = struct.unpack_from(e + "h", buf, p)[0]
            return None if v == 0x7FFF else v
        if b in (0x04, 0x0B):
            v = struct.unpack_from(e + "H", buf, p)[0]
            return None if v == 0xFFFF else v
        if b == 0x05:
            v = struct.unpack_from(e + "i", buf, p)[0]
            return None if v == 0x7FFFFFFF else v
        if b in (0x06, 0x0C):
            v = struct.unpack_from(e + "I", buf, p)[0]
            return None if v == 0xFFFFFFFF else v
        if b == 0x08:
            return struct.unpack_from(e + "f", buf, p)[0]
        if b == 0x09:
            return struct.unpack_from(e + "d", buf, p)[0]
        return None  # strings / 64-bit ignored

    return read


def parse_fit(buf):
    if len(buf) < 14:
        raise ValueError("FIT muito curto")
    header_size = buf[0]
    data_size = struct.unpack_from("<I", buf, 4)[0]
    if buf[8:12] != b".FIT":
        raise ValueError("no .FIT")
    end = min(header_size + data_size, len(buf))
    pos = header_size
    defs = {}
    records = []
    last_ts = None   # running timestamp for compressed-timestamp headers (5-bit offset, 32 s rollover)
    sport = None     # first sport enum seen (msg 12 field 0, or session msg 18 field 5)
    while pos < end:
        rh = buf[pos]
        pos += 1
        ts_offset = None
        is_def = has_dev = False
        if rh & 0x80:
            local = (rh >> 5) & 0x03
            ts_offset = rh & 0x1F
        else:
            local = rh & 0x0F
            is_def = bool(rh & 0x40)
            has_dev = bool(rh & 0x20)
        if is_def:
            pos += 1
            little = buf[pos] == 0
            pos += 1
            gmn = struct.unpack_from("<H" if little else ">H", buf, pos)[0]
            pos += 2
            nf = buf[pos]
            pos += 1
            fields = []
            for _ in range(nf):
                fields.append((buf[pos], buf[pos + 1], buf[pos + 2]))  # num, size, bt
                pos += 3
            dev_size = 0
            if has_dev:
                nd = buf[pos]
                pos += 1
                for _ in range(nd):
                    dev_size += buf[pos + 1]
                    pos += 3
            defs[local] = {"gmn": gmn, "little": little, "fields": fields,
                           "devSize": dev_size, "read": _reader(buf, little)}
        else:
            d = defs.get(local)
            if d is None:
                raise ValueError("FIT corrompido (dado sem definição)")
            p = pos
            rec = {}
            read = d["read"]
            gmn = d["gmn"]
            for num, size, bt in d["fields"]:
                if gmn == 20:
                    v = read(p, bt)
                    if v is not None:
                        if num == 0:
                            rec["lat"] = v * (180 / 2147483648)
                        elif num == 1:
                            rec["lon"] = v * (180 / 2147483648)
                        elif num == 2:
                            if "alt" not in rec:
                                rec["alt"] = v / 5 - 500
                        elif num == 78:
                            rec["alt"] = v / 5 - 500
                        elif num == 5:
                            rec["dist"] = v / 100
                        elif num == 6:
                            if "speed" not in rec:
                                rec["speed"] = v / 1000
                        elif num == 73:
                            rec["speed"] = v / 1000
                        elif num == 7:
                            rec["power"] = v
                        elif num == 4:
                            rec["cad"] = v          # cadence (rpm) — 0 ⇒ not pedalling
                        elif num == 253:
                            rec["time"] = v
                elif (gmn == 12 and num == 0) or (gmn == 18 and num == 5):
                    v = read(p, bt)
                    if v is not None and sport is None:
                        sport = v
                elif num == 253:   # any message's timestamp advances the running clock
                    v = read(p, bt)
                    if v is not None:
                        rec["time"] = v
                p += size
            pos = p + d["devSize"]
            # compressed-timestamp header: reconstruct the time from the 5-bit offset
            if ts_offset is not None and "time" not in rec and last_ts is not None:
                ts = (last_ts & ~31) | ts_offset
                if ts < last_ts:
                    ts += 32
                rec["time"] = ts
            if "time" in rec:
                last_ts = rec["time"]
            if gmn == 20:
                records.append(rec)
    return records, sport


FIT_EPOCH = 631065600   # 1989-12-31 UTC, seconds


# ---- JSON.stringify(out, null, 1) — byte-compatible writer ----
# json.dumps differs: it escapes non-ASCII (ensure_ascii) and renders
# integer-valued floats with a trailing .0 (JS: String(2.0) === '2').

_ESC = {'"': '\\"', "\\": "\\\\", "\b": "\\b", "\f": "\\f",
        "\n": "\\n", "\r": "\\r", "\t": "\\t"}


def jquote(s):
    out = ['"']
    for ch in s:
        if ch in _ESC:
            out.append(_ESC[ch])
        elif ch < " ":
            out.append("\\u%04x" % ord(ch))
        else:
            out.append(ch)
    out.append('"')
    return "".join(out)


def jstringify(v, ind=""):
    if v is None:
        return "null"
    if isinstance(v, bool):
        return "true" if v else "false"
    if isinstance(v, str):
        return jquote(v)
    if isinstance(v, int):
        return js_str(v)
    if isinstance(v, float):
        return js_str(v) if math.isfinite(v) else "null"   # JSON.stringify(NaN) === 'null'
    ni = ind + " "
    if isinstance(v, list):
        if not v:
            return "[]"
        return "[\n" + ",\n".join(ni + jstringify(x, ni) for x in v) + "\n" + ind + "]"
    if not v:
        return "{}"
    return ("{\n"
            + ",\n".join(ni + jquote(k) + ": " + jstringify(val, ni) for k, val in v.items())
            + "\n" + ind + "}")


# ---- scan ----

files = sorted(f for f in os.listdir(DIR) if f.endswith(".fit") or f.endswith(".fit.gz"))
out = []
errors = 0
for f in files:
    try:
        with open(os.path.join(DIR, f), "rb") as fh:
            buf = fh.read()
        if f.endswith(".gz"):
            buf = gzip.decompress(buf)
        recs, sport = parse_fit(buf)
        if len(recs) < 2:
            raise ValueError("sem registros")
        n = len(recs)
        nPow = nPowPos = nCad = nAlt = 0
        maxDist = 0
        t0 = t1 = None
        for r in recs:
            if r.get("power") is not None:
                nPow += 1
                if r["power"] > 0:
                    nPowPos += 1
            if r.get("cad") is not None:
                nCad += 1
            if r.get("alt") is not None:
                nAlt += 1
            if r.get("dist") is not None and r["dist"] > maxDist:
                maxDist = r["dist"]
            if r.get("time") is not None:
                if t0 is None:
                    t0 = r["time"]
                t1 = r["time"]
        sport_name = SPORT.get(sport)
        if sport_name is None:
            sport_name = "unknown" if sport is None else "sport" + js_str(sport)
        out.append({
            "id": re.sub(r"\.fit(\.gz)?$", "", f),
            "file": os.path.join("strava_ppaz", f),
            "sport": sport_name,
            "date": (datetime.fromtimestamp(t0 + FIT_EPOCH, timezone.utc).strftime("%Y-%m-%d")
                     if t0 is not None else None),
            "hours": float(to_fixed((t1 - t0) / 3600, 2)) if t0 is not None else None,
            "km": float(to_fixed(maxDist / 1000, 1)),
            "n": n,
            "powCov": float(to_fixed(nPow / n, 3)),
            "powPos": float(to_fixed(nPowPos / n, 3)),
            "cadCov": float(to_fixed(nCad / n, 3)),
            "altCov": float(to_fixed(nAlt / n, 3)),
        })
    except Exception:
        errors += 1
with open(os.path.join(DATA, "strava_ppaz_manifest.json"), "w", encoding="utf-8") as fh:
    fh.write(jstringify(out))

# ---- summary ----
bySport = {}
for a in out:
    bySport.setdefault(a["sport"], []).append(a)
print(f"P. PAZ STRAVA EXPORT — {len(files)} FIT files, {len(out)} parsed, {errors} errors\n")
print("sport      n     w/ power(>50% cov)   rides>20km w/ power")
for s, arr in sorted(bySport.items(), key=lambda kv: len(kv[1]), reverse=True):
    pow_ = [a for a in arr if a["powCov"] > 0.5]
    big = [a for a in pow_ if a["km"] > 20]
    print(s.ljust(9) + str(len(arr)).rjust(5) + str(len(pow_)).rjust(13) + str(len(big)).rjust(19))
rides = [a for a in bySport.get("ride", []) if a["powCov"] > 0.5]
if rides:
    dates = sorted(r["date"] for r in rides if r["date"])
    kms = sorted(r["km"] for r in rides)
    q = lambda p: kms[math.floor(p * (len(kms) - 1))]
    print(f"\nRIDES WITH POWER: {len(rides)}")
    print(f"  dates {dates[0]} … {dates[-1]}")
    print(f"  km: min {js_str(q(0))}  p25 {js_str(q(.25))}  median {js_str(q(.5))}  "
          f"p75 {js_str(q(.75))}  max {js_str(q(1))}")
    print(f"  alt coverage ≥99%: {len([r for r in rides if r['altCov'] >= 0.99])}")
    print(f"  cadence coverage ≥50%: {len([r for r in rides if r['cadCov'] >= 0.5])}")
print("\nwrote strava_ppaz_manifest.json")
